javalib-ex/src/main/scala/java/io/DataInputStream.scala

```diff
@@ -69,24 +54,27 @@ class DataInputStream(in: InputStream) extends FilterInputStream(in)
   }
 
   def readChar(): Char = {
-    if (hasArrayBuffer)
+    if (hasArrayBuffer) {
       bufDataView.getUint16(consumePos(2)).toChar
-    else
-      view(2).getUint16(0).toChar
+    } else {
+      ((readByte() << 8) | (readByte() & 0xff)).toChar
+    }
   }
 
   def readDouble(): Double = {
-    if (hasArrayBuffer)
+    if (hasArrayBuffer) {
       bufDataView.getFloat64(consumePos(8))
-    else
-      view(8).getFloat64(0)
+    } else {
+      java.lang.Double.longBitsToDouble(readLong())
```

Comment:
Calling `readLong()` will do the `if (hasArrayBuffer)` again. Is it intentional? Should we move the relevant part of `readLong()` in a `private def` instead?